[Refactor] Add infra-driver, cluster-runtime, container-runtime, registry-configurator interface

justadogistaken opened the pull request:
<!--  Thanks for submitting a pull request! Here are some tips for you:
1. Please make sure you have read and understood the contributing guidelines: https://github.com/alibaba/blob/master/CONTRIBUTING.md
2. Please make sure the PR has a corresponding issue.
-->

### Describe what this PR does / why we need it


### Does this pull request fix one issue?

<!--If that, add "Fixes #xxxx" below in the next line. For example, Fixes #15. Otherwise, add "NONE" -->

### Describe how you did it


### Describe how to verify it


### Special notes for reviews

VinceCui committed:
Stash

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

VinceCui committed:
Stash

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

VinceCui committed:
Add infra-driver, cluster-runtime, container-runtime, registry-configurator interface;
Add main process in Install() and ScaleUp().

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

VinceCui committed:
Add infra-driver, cluster-runtime, container-runtime, registry-configurator interface;
Add main process in Install() and ScaleUp().

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

justadogistaken commented:
@VinceCui FYI

kakaZhou719 commented on pkg/container-runtime/installer.go:
```diff
@@ -0,0 +1,42 @@
+
+type Config struct {
+	Type         string
+	LimitNofile  string `json:"limitNofile,omitempty" yaml:"limitNofile,omitempty"`
```
should this field be `int` type? and  Is this field for checking OS configuration?

kakaZhou719 replied:
should this field be `int` type?

VinceCui committed:
Add vendor.

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

VinceCui committed:
Merge remote-tracking branch 'upstream/main' into feature/reconstruct-run

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

# Conflicts:
#	cmd/sealer/cmd/run.go

kakaZhou719 committed:
add ssh infra driver (#1622)

* Add infra-driver, cluster-runtime, container-runtime, registry-configurator interface;
Add main process in Install() and ScaleUp().

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

* add ssh infra driver

* update reviews

* add cluster infra info

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>
Co-authored-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

VinceCui committed:
Merge remote-tracking branch 'upstream/feature/reconstruct-run' into feature/reconstruct-run

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

# Conflicts:
#	cmd/sealer/cmd/run.go
#	pkg/infradriver/infradriver.go
#	pkg/infradriver/ssh_infradriver.go
#	pkg/runtime/interface.go
#	vendor/github.com/spf13/cobra/bash_completionsV2.go
#	vendor/github.com/spf13/cobra/user_guide.md
#	vendor/github.com/spf13/viper/watch.go

Stevent-fei committed:
Add Container runtime---docker install (#1652)

* modify docker install

* modify error message

* modify review,add containerdInstaller.go file,add error return

* add docker info

* Docker installation uses docker.sh, docker uninstall uses docker-uninstall.sh

* Optimize error messages and custom variables

kakaZhou719 committed:
feat:add registry and container-runtime configurator (#1642)

* feat:add registry

* feat:add container runtime configurator

* modify registry part

* add registry config part

* update reviews

* update reviews

VinceCui committed:
Merge remote-tracking branch 'upstream/feature/reconstruct-run' into feature/reconstruct-run

Stevent-fei committed:
optimization: optimization error message (#1664)

* optimization: optimization error message

* add docker cert dir

VinceCui committed:
Reconstruct kube-runtime.

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

VinceCui committed:
Merge remote-tracking branch 'upstream/feature/reconstruct-run' into feature/reconstruct-run

VinceCui committed:
Merge branch 'feature/reconstruct-run' into feature/kube-runtime

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

# Conflicts:
#	pkg/container-runtime/docker.go

kakaZhou719 committed:
Add installer hooks  (#1669)

* add hook

* if on field is not set, then use all host ip as default

update reviews

fix bug:Make sure each host got from Scope is in the given host ip list.

* fix bug:Make sure each host got from Scope is in the given host ip list.

kakaZhou719 committed:
Supplement the installation processes in k8s runtime (#1682)

* add comment

* add runtime driver

add function comment

* update reviews

update reviews

* update reviews

VinceCui committed:
Merge remote-tracking branch 'upstream/main' into feature/reconstruct-run

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

# Conflicts:
#	apply/processor/scale.go
#	cmd/sealer/cmd/cluster/delete.go
#	cmd/sealer/cmd/cluster/join.go
#	cmd/sealer/cmd/cluster/run.go
#	pkg/clusterfile/clusterfile.go
#	pkg/clusterfile/pre_process.go
#	pkg/image/default_image.go
#	pkg/image/image.go
#	pkg/image/image_interface.go
#	pkg/runtime/kubernetes/join_masters.go
#	pkg/runtime/kubernetes/join_nodes.go
#	pkg/runtime/kubernetes/registry.go
#	pkg/runtime/kubernetes/utils.go
#	vendor/modules.txt

kakaZhou719 committed:
add run and join (#1675)

* add run cmd and join cmd

* update reviews

* upload registry data

* add host env map for infradriver

* update reviews

VinceCui committed:
Merge branch 'feature/reconstruct-run' of https://github.com/alibaba/sealer into feature/reconstruct-run

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

# Conflicts:
#	pkg/config/config.go

kakaZhou719 committed:
chore: fix code lint (#1696)

Stevent-fei committed:
Add run join delete (#1699)

* add run join delete support

Optimize run join and delete

* Optimize

* modify join and delete

* modify join and delete

* Modify the content of the review

Stevent-fei committed:
fix saveAll funcation,add unit test (#1709)

kakaZhou719 committed:
delete unused code (#1706)

* save commit

* chore:delete unused code

* rebase code

* rebase code

allencloud commented:
@sealerio/sealer-reviewers 
I still have several questions for this refactoring. I must say again that I am the one who totally support refactoring sealer's arch. While to achieve this goal, I wish the refactoring team would stay serious to face the challenges, back to basic.

1. I see this pull request already contains two much code change without maintainers' review, maybe contains reviews just from @kakaZhou719 and @VinceCui . Then, how to guarantee that this pull request could be rebased to the latest main branch? The rebase work including solving conflict would consume so much energy of the entire team.
2. How to detect whether the rebase work would ignore the original bugfix code by accident? The refactoring will indeed change the architecture, but for the detailed functionality, we have no way to notice the side effect.
3. No testing is guaranteed, how to prove the quality of refactoring? Are we changing the legacy bad world into a new bad world.  How to reduce the possibility that we refatored the arch into another unsuccessful arch.

In a word, if there is no a reasonably detailed way for this pull request to guarantee merging to the main branch, the refactoring seems to help less.

Commit:
bugfix for sealer run

Commit:
add env and config

VinceCui committed:
Merge remote-tracking branch 'upstream/main' into feature/reconstruct-run

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

# Conflicts:
#	apply/driver/local.go
#	apply/processor/create.go
#	apply/processor/delete.go
#	apply/processor/gen.go
#	apply/processor/scale.go
#	cmd/sealer/cmd/cluster/apply.go
#	cmd/sealer/cmd/cluster/delete.go
#	cmd/sealer/cmd/utils/utils.go
#	pkg/runtime/kubernetes/runtime.go

VinceCui committed:
Merge branch 'feature/reconstruct-run' of https://github.com/alibaba/sealer into feature/reconstruct-run

Commit:
remove plugin dir

Commit:
merge code

Commit:
add config to join and delete

VinceCui committed:
Merge branch 'feature/reconstruct-run' of https://github.com/alibaba/sealer into feature/reconstruct-run

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

# Conflicts:
#	pkg/plugin/utils.go

VinceCui committed:
Merge remote-tracking branch 'upstream/main' into feature/reconstruct-run

Signed-off-by: huaiyou <huaiyou.cyz@alibaba-inc.com>

# Conflicts:
#	apply/driver/local.go
#	apply/processor/create.go
#	apply/processor/gen.go

Commit:
fixbug: send cluster cert to master0

Commit:
bugfix:run,join,delete main process

Commit:
delete unused func

Commit:
add unit test for infra,env,config,hooks

Commit:
fix lint

Commit:
bugfix: run confirm logic first before pre-uninstall-hook;bugix:return error if get node name failed

justadogistaken committed:
fix ci lint

Signed-off-by: David Bao <baojn1998@163.com>

kakaZhou719 committed:
code optimization for sealer CLI (#1727)

* polish CLI: delete unused flag, add exmaples for join,delete,run;let all master could access api server

* fix lint

fix lint

fix lint

* fix lint

Commit:
fix ci: golint issue

Commit:
fix ci: golint issue

Commit:
optimize: make docker installer logic more clear

Stevent-fei committed:
Save clusterfile (#1736)

* Save clusterfile

* modify clusterfile ut

Commit:
bugfix:write config if path not exist;joined master can access api server

Commit:
merge code

merge code

Commit:
merge code

Commit:
merge code from main

Commit:
merge code from main

Commit:
merge code from main

VinceCui commented on pkg/runtime/kubernetes/runtime.go:
```diff
@@ -15,390 +15,244 @@
+			KubeadmConfigFromClusterFile: clusterFileKubeConfig,
+			APIServerDomain:              DefaultAPIserverDomain,
+			//TODO
+			LvsImage:             fmt.Sprintf("%s/fanux/lvscare:latest", registryInfo.URL),
```
use labring/lvscare:v1.1.3-beta.8

VinceCui commented on pkg/runtime/kubernetes/init.go:
```diff
@@ -17,170 +17,110 @@ package kubernetes
-	k.setControlPlaneEndpoint(fmt.Sprintf("%s:6443", k.getAPIServerDomain()))
-	if k.APIServer.ExtraArgs == nil {
-		k.APIServer.ExtraArgs = make(map[string]string)
+	//TODO, save it into kubernetes
```
no need do this TODO, kubeadm will do it by itself.

VinceCui reviewed (commented):
There are so many reviews so far, I will continue to add

Commit:
merge code from main

Commit:
fix reviews

Commit:
fix reviews for reconstruct

Commit:
merge code from main

Commit:
tidy project vendor

Commit:
fix FOSSA issue

Commit:
fix FOSSA issue
